internal/service/sso/data_source_application.go

```diff
@@ -0,0 +1,1166 @@
+		)
+
+		return
+	}
+
+	r.Client = resourceConfig.Client.API
+	if r.Client == nil {
+		resp.Diagnostics.AddError(
+			"Client not initialised",
+			"Expected the PingOne client, got nil.  Please report this issue to the provider maintainers.",
+		)
+		return
+	}
+}
+
+func (r *ApplicationDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data *applicationDataSourceModel
+
+	if r.Client.ManagementAPIClient == nil {
+		resp.Diagnostics.AddError(
+			"Client not initialized",
+			"Expected the PingOne client, got nil.  Please report this issue to the provider maintainers.")
+		return
+	}
+
+	// Read Terraform prior state data into the model
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	var application management.ReadOneApplication200Response
+
+	// Application API does not support SCIM filtering
+	if !data.ApplicationId.IsNull() {
+		// Run the API call
+		var response *management.ReadOneApplication200Response
+		resp.Diagnostics.Append(framework.ParseResponse(
+			ctx,
+
+			func() (any, *http.Response, error) {
+				return r.Client.ManagementAPIClient.ApplicationsApi.ReadOneApplication(ctx, data.EnvironmentId.ValueString(), data.ApplicationId.ValueString()).Execute()
+			},
+			"ReadOneApplication",
+			framework.DefaultCustomError,
+			sdk.DefaultCreateReadRetryable,
+			&response,
+		)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+
+		application = *response
+
+	} else if !data.Name.IsNull() {
+		// Run the API call
+		var entityArray *management.EntityArray
+		resp.Diagnostics.Append(framework.ParseResponse(
+			ctx,
+
+			func() (any, *http.Response, error) {
+				return r.Client.ManagementAPIClient.ApplicationsApi.ReadAllApplications(ctx, data.EnvironmentId.ValueString()).Execute()
+			},
+			"ReadAllApplications",
+			framework.DefaultCustomError,
+			sdk.DefaultCreateReadRetryable,
+			&entityArray,
+		)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+
+		if applications, ok := entityArray.Embedded.GetApplicationsOk(); ok {
+			found := false
+
+			for _, applicationObject := range applications {
+				applicationInstance := applicationObject.GetActualInstance()
+
+				applicationName := ""
+
+				switch v := applicationInstance.(type) {
+				case *management.ApplicationExternalLink:
+					applicationName = v.GetName()
+
+				case *management.ApplicationOIDC:
+					applicationName = v.GetName()
+
+				case *management.ApplicationSAML:
+					applicationName = v.GetName()
+				}
+
+				if applicationName == data.Name.ValueString() {
+					application = applicationObject
+					found = true
+					break
+				}
+			}
+
+			if !found {
+				resp.Diagnostics.AddError(
+					"Cannot find the application from name",
+					fmt.Sprintf("The application name %s for environment %s cannot be found", data.Name.String(), data.EnvironmentId.String()),
```

Comment:
It might be worth adding to this that the data source can only retrieve saml, oidc and external link applications